### DISCUSSION TITLE
关于Seata服务器支持更多SPI扩展点的讨论

### DISCUSSION THREAD

**Discussion by HEIBAIJIE** [apache/incubator-seata#5156] [Ideas]
我们发现了一些可能有益的新的SPI扩展点：
1. 消息类型、编解码器扩展：使Seata服务器支持对自定义消息类型的编解码
2. 消息处理钩子的扩展：在ServerOnRequestProcessor处理某类消息前后实现额外的扩展逻辑
3. 处理器扩展：允许服务器为上述自定义消息类型注册自定义处理器
4. 启动项扩展：允许Seata服务器在启动时执行额外的逻辑
5. 具体配置扩展：通过简单的实现来增加一个具体的配置项
6. 客户端相关：对原有硬编码进行少量变动，使Seata能与RM、TM以外的客户端类型建立连接

目前我们的功能已经基本实现，但我们并不奢望直接将具体功能合入，或者直接合入我们对扩展点的实现。
只是想在此了解一下社区对相关SPI扩展点的看法，讨论未来支持这些扩展点的可能性，欢迎给出意见和建议。

**Comment by funky-eyes** [comment]
1. 必然需要修改seata-server侧的代码，所以是直接侵入的。
2. 这块就是rm侧的hook就可以完成， 计划中
3. 不应该让业务参与到消息处理器的程度来，据我了解没有什么组件这么干的，你可以举几个别的主流主键支持把自己的processor开放出来允许业务随意注册的，既然注册肯定是有地方要魔改代码否则怎么收发消息rpc？
4.Seata侧考虑做一些spi的方式，现在已经通过loader.path将libs下的所有jar装载，那么只要spi的实现类的jar被放入到这个路径下就能被感知到
5.不明白什么意思
6.为什么要跟rm tm以外的客户端建立连接？kafka，rocketmq这类有server的消息中间件有这么干的？再producer和consumer外允许其他非标连接？连接的意义是什么？

**Reply by HEIBAIJIE** [reply]
@a364176773 
关于Server侧的扩展点增强，我们又思考了一下，3、5、6可能确实没有必要。但是其他扩展点我想再解释一下：
1. 关于Server的启动扩展
我们有一个诉求是通过SPI的方式，在服务器启动后、关机前执行额外的扩展逻辑。
这样的需求在Dubbo中可以通过Container扩展点来实现，通过实现Container接口就可以实现启动时执行start()逻辑，终止前执行end()逻辑。但是在Seata Server中暂时还没有这样的机制。

2. 关于rpcHook的扩展
目前在AbstractNettyRemoting中存在RpcHook机制，在发送消息前、后可以执行自定义逻辑。但是在processMessage(...)处理消息的过程中缺乏同样的Hook机制，导致无法在消息发送后、接受前做一些对称处理。
可以类比RpcHook，增加ProcessHook。在调用`pair.getFirst().process(ctx, rpcMessage);`之前、之后，分别调用doBeforeProcessHooks和doAfterProcessHooks。
参考Dubbo的Filter机制，也是可以在消息发送、处理的前后执行自定义逻辑的。

3. 关于消息类型的扩展
我们有扩展Seata消息类型的诉求。消息的生成和发送目前可以通过对Core的SPI扩展来实现，DefaultCoordinator也可以很好地处理toTC类型的消息。但目前发现，消息流程中有两个地方是硬编码的，导致扩展比较困难。
第一个是消息的编解码。MessageCodecFactory中通过一个静态方法Switch消息码的方式来获取编解码器。可否考虑提高这部分实现的扩展性？
第二个是消息在处理器中的注册。魔改处理器本身可能不太合适，但是可否考虑允许新的消息类型在OnRequestProcessor上注册？
如果这两处硬编码可以解耦，那么则Seata消息机制的可扩展性会提高很多。

**Reply by funky-eyes** [reply]
@HEIBAIJIE 
1.dubbo是可以直接修改provider侧代码，因为他是集成到业务sdk中的，而seata的provider也就是seata-server必然存在修改源码的情况，大部分扩展的代码也要在源码基础上做实现和打包，在这种情况下貌似业务方都可以直接部署一个被其修改过源码的seata-server了，就感觉spi不是很必要了
2.第二点我觉得你是不是想做rm侧的钩子，当tc下发二阶段请求的时候业务可以做一些动作？
3.可以展开说说去除MessageCodecFactory的Switch方式后，什么设计模式运用上会好一些？还有这块可以不使用默认的seata序列化，可以用kryo或fst之类的，后续可能考虑扩展fastjson2的jsonb方式，其次是在server上考虑新的消息类型是干啥作用我还是没明白

**Reply by HEIBAIJIE** [reply]
@a364176773 
感谢您的回复，我再解释一下。
1. 扩展代码确实需要依赖Seata的jar包进行编译，但是可以独立仓库、独立编译。
若有扩展接口，那么引入扩展时就无需对Seata-server本体的源码或jar包进行任何修改。只是部署时在/lib目录额外放一个扩展jar包。
这样就将本体-扩展的双向依赖降为扩展对本体的单向依赖，在跟随社区版本升级的时候减少一些痛苦。
当然，对seata-server的定制扩展和本体-扩展解耦的需求，可能是一个相对小众的需求。如果社区有顾虑我们也理解。
2. 我们希望做的是seata-server侧的消息处理钩子，虽然这个钩子RM侧也可以使用。如果目的是RM侧的消息处理钩子，那么TransactionHook已经提供非常强大的能力了。
3. 之前确实没有考虑过使用kryo或fst序列化，谢谢提醒。
不用switch的话，可以考虑用map重构原有的简单工厂模式，参考[https://segmentfault.com/a/1190000021641277](url)。
我们考虑新的消息类型，是因为在我们的架构中需要seata-server之间通信，进行一些事务消息的封装和转发。其他人有可能会有其他的需求。

**Reply by funky-eyes** [reply]
1.大概明白了
2.可以举一些server上消息处理钩子的一些用途吗？做旁路分析采集，或者metric？
3.我想了解下什么样的架构需要与seata-server进行通信呢？如果已经有了第二点的支持，还需要3吗？

**Reply by HEIBAIJIE** [reply]
@a364176773
1. 关于消息处理钩子，我们目前用到的场景是拦截RM与server的交互，对上下文中的resourceID在Server侧进行一个动态的扩展。
RM注册、上行时为resourceID添加一些标识，便于下行时找到特定范围的RM。
标识的内容只有Server侧才配置，业务侧对此不感知。
当然消息处理钩子还是有很多其他的扩展潜力的，比如您提到的旁路分析，特定的日志打印等等。
2. 根据我们的压测，在db模式、超高吞吐量的条件下，Seata服务器的性能表现严重受制于数据库交互，而且写多于读。所以我们让不同的Server集群连接不同的数据库实例，让绝大多数局域的分布式事务需求在该局域内解决，这样就把压力从一个数据库实例分摊到了多个实例。但是依旧存在一些跨域的事务，显然按照Seata现有的架构，如果全局事务与分支事务不在同一个数据库中，分支是无法注册成功的。那么我们就需要server之间通信，把分支事务转到全局事务所在的server处理。
这个只有钩子是无法支持，一个简单的例子，seata-server原本不是用来处理分支提交/回滚的，但是某些时候我们需要它中转这种消息。

**Reply by funky-eyes** [reply]
明白了，第二点我建议直接用现有的分布式数据库比如oceanbase或者tidb来做分库分表，对seata-server而言还是连得逻辑表和逻辑库即可，你方便添加的们的钉钉群吗，我们下次月度讨论可以带上你这个需求，如果你有空的话可以跟我们做个在线分享或交流，让社区其他的开发者更清晰直接的了解这个需求。我们开发者钉钉群为33774360

**Reply by funky-eyes** [reply]
其次我们在2.0上会推出raft模式，目前已经是在跑各种压力测试和故障测试中，后续还会推出multi-raft模式，将不同的事务随机路由到不同的raft group中，比如3台tc，可以搭建3个group，那么每台为一个group的leader，其余两台为group的副本，来提升吞吐量，我们的测试结果相对db和redis模式下还是提升非常大的，如果你有兴趣也可以参与到这个pr的review和测试中 https://github.com/seata/seata/pull/5226

**Reply by HEIBAIJIE** [reply]
非常感谢，已加入钉钉群。

**Comment by funky-eyes** [comment]
通过社区内部沟通,seata将会考虑以下扩展点
1.消息处理钩子的扩展：在ServerOnRequestProcessor处理某类消息前后实现额外的扩展逻辑
2.处理器扩展：允许服务器为上述自定义消息类型注册自定义处理器
